plugins/rts54hub/fu-rts54hub-rtd21xx-mergeinfo.c

```diff
@@ -0,0 +1,481 @@
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_mergeinfo_ensure_version(FuRts54hubRtd21xxMergeinfo *self, GError **error)
+{
+	guint8 buf_reply[16] = {0x00};
+	guint8 buf_request[2] = {0x00};
+	g_autofree gchar *version = NULL;
+
+	/* read merge version */
+	buf_request[0] = DDCCI_COMM_FIRST_OPCODE;
+	buf_request[1] = DDCCI_GET_VERSION_OPCODE;
+	if (!fu_rts54hub_rtd21xx_device_ddcci_write(FU_RTS54HUB_RTD21XX_DEVICE(self),
+						    DDCCI_TARGET_ADDR,
+						    DDCCI_COMM_SUB_ADDR,
+						    buf_request,
+						    sizeof(buf_request),
+						    error)) {
+		g_prefix_error(error, "failed to DDC/CI communication with fw: ");
+		return FALSE;
+	}
+
+	/* wait for device ready */
+	fu_device_sleep(FU_DEVICE(self), 300); /* ms */
+
+	if (!fu_rts54hub_rtd21xx_device_ddcci_read(FU_RTS54HUB_RTD21XX_DEVICE(self),
+						   DDCCI_TARGET_ADDR,
+						   DDCCI_COMM_SUB_ADDR,
+						   buf_reply,
+						   sizeof(buf_reply),
+						   error)) {
+		g_prefix_error(error, "failed to DDC/CI communication with fw: ");
+		return FALSE;
+	}
+
+	/* set merge version */
+	fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_QUAD);
+	version =
+	    g_strdup_printf("%u.%u.%u.%u", buf_reply[4], buf_reply[5], buf_reply[6], buf_reply[7]);
+	fu_device_set_version(FU_DEVICE(self), version);
+
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_mergeinfo_read_version(FuRts54hubRtd21xxMergeinfo *self,
+					   guint8 *buf_version,
+					   GError **error)
+{
+	guint8 buf_reply[16] = {0x00};
+	guint8 buf_request[2] = {0x00};
+
+	/* read merge version */
+	buf_request[0] = DDCCI_COMM_FIRST_OPCODE;
+	buf_request[1] = DDCCI_GET_VERSION_OPCODE;
+	if (!fu_rts54hub_rtd21xx_device_ddcci_write(FU_RTS54HUB_RTD21XX_DEVICE(self),
+						    DDCCI_TARGET_ADDR,
+						    DDCCI_COMM_SUB_ADDR,
+						    buf_request,
+						    sizeof(buf_request),
+						    error)) {
+		g_prefix_error(error, "failed to DDC/CI communication with fw: ");
+		return FALSE;
+	}
+
+	/* wait for device ready */
+	fu_device_sleep(FU_DEVICE(self), 300); /* ms */
+
+	if (!fu_rts54hub_rtd21xx_device_ddcci_read(FU_RTS54HUB_RTD21XX_DEVICE(self),
+						   DDCCI_TARGET_ADDR,
+						   DDCCI_COMM_SUB_ADDR,
+						   buf_reply,
+						   sizeof(buf_reply),
+						   error)) {
+		g_prefix_error(error, "failed to DDC/CI communication with fw: ");
+		return FALSE;
+	}
+
+	if (!fu_memcpy_safe(buf_version,
+			    VERSION_NUMBER_COUNT,
+			    0x0, /* dst */
+			    (const guint8 *)buf_reply,
+			    16,
+			    4, /* src */
+			    VERSION_NUMBER_COUNT,
+			    error)) {
+		g_prefix_error(error, "memcpy merge version fail: ");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_mergeinfo_write_version(FuRts54hubRtd21xxMergeinfo *self,
+					    const guint8 *buf_version,
+					    gsize buf_size,
+					    GError **error)
+{
+	guint8 buf_request[6] = {0x00};
+
+	if (buf_size != VERSION_NUMBER_COUNT) {
+		g_prefix_error(error, "failed to check version buffer size: ");
```

Comment:
g_prefix_error is for when you're adding to an existing GError -- you need `g_set_error()` here